src/DataStructures/Tensor/EagerMath/Magnitude.hpp

```diff
@@ -54,35 +54,51 @@ Scalar<DataType> sqrt_magnitude(const Scalar<DataType>& input) noexcept {
   return Scalar<DataType>{sqrt(abs(get(input)))};
 }
 
-/// @{
 /// \ingroup TensorGroup
-/// \brief Compute $\\ell^2$ (Frobenius) norm of a tensor
+/// \brief Compute point-wise $\\ell^2$ (Frobenius) norm of a tensor
 ///
 /// \details
-/// Computes the Frobenius norm of a given tensor of arbitrary rank.
+/// Computes the point-wise Frobenius norm of a given tensor of arbitrary rank.
 template <typename DataType, typename Symm, typename IndexList>
-Scalar<DataType> l2_norm(
+Scalar<DataType> pointwise_l2_norm(
     const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
-  auto& input =
-      const_cast<Tensor<DataType, Symm, IndexList>&>(tensor);  // NOLINT
-  DataType sum(tensor.begin()->size(), 0.);
-  for (auto it = input.begin(); it != input.end(); ++it) {
-    sum += input.multiplicity(it) * square(*it);
+  DataType pointwise_l2_normsq{tensor.begin()->size(), 0.};
+  for (auto& tensor_element : tensor) {
+    // In order to handle tensory symmetries, we multiply the square of each
+    // stored component with its multiplicity
+    pointwise_l2_normsq +=
+        tensor.multiplicity(const_cast<DataType*>(&tensor_element)) *  // NOLINT
+        square(tensor_element);
+    // `const_cast` above is necessary to stop gcc warnings (errors) about
+    // the implicit conversion of function argument from a pointer to const to
+    // pointer to non-const type.
   }
-  return Scalar<DataType>{sqrt(sum)};
+  return Scalar<DataType>{sqrt(pointwise_l2_normsq)};
 }
+
+/// \ingroup TensorGroup
+/// \brief Compute $\\ell^2$ (Frobenius) norm of a tensor
+///
+/// \details
+/// Computes the Frobenius norm of a given tensor of arbitrary rank,
+/// averaged over all grid points.
 template <typename DataType, typename Symm, typename IndexList>
 double l2_norm(const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
-  auto& input =
-      const_cast<Tensor<DataType, Symm, IndexList>&>(tensor);  // NOLINT
-  DataType sum(tensor.begin()->size(), 0.);
-  for (auto it = input.begin(); it != input.end(); ++it) {
-    sum += input.multiplicity(it) * square(*it);
+  DataType pointwise_l2_normsq{tensor.begin()->size(), 0.};
```

Comment:
how about you move the shared code between these two functions into a `detail::pointwise_l2_norm_square` function?

Reply:
done